virtualenv setup failure during install test on Python 3.7 with minimum package levels

The Travis run on OS-X on Python 3.7.7 with minimum package levels fails when setting up the virtualenv for an install test (see https://travis-ci.org/github/pywbem/pywbem/jobs/674559955):

```
Testcase test1: Pip install from repo root directory: tests/installtest/../..
Saving current virtualenv: /Users/travis/venv
Creating virtualenv: tests/installtest/../../tmp_installtest/virtualenvs/pywbem_test_test1
..using this Python command: /Users/travis/venv/bin/python

Error: Command failed with rc=100: virtualenv -p /Users/travis/venv/bin/python tests/installtest/../../tmp_installtest/virtualenvs/pywbem_test_test1, output follows:

/Users/travis/venv/lib/python3.7/site-packages/virtualenv.py:991: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
  import imp
Using base prefix '/usr/local/opt/python/bin/../Frameworks/Python.framework/Versions/3.7'
New python executable in /Users/travis/build/pywbem/pywbem/tmp_installtest/virtualenvs/pywbem_test_test1/bin/python
ERROR: The executable /Users/travis/build/pywbem/pywbem/tmp_installtest/virtualenvs/pywbem_test_test1/bin/python is not functioning
ERROR: It thinks sys.prefix is '/usr/local/Cellar/python/3.7.7/Frameworks/Python.framework/Versions/3.7' (should be '/Users/travis/build/pywbem/pywbem/tmp_installtest/virtualenvs/pywbem_test_test1')
ERROR: virtualenv is not compatible with this system or executable
Running virtualenv with interpreter /Users/travis/venv/bin/python
make: *** [installtest] Error 1
```

Since this was a run with minimum package levels, it used virtualenv 14.0.0. That version does not state support for Python 3.7 yet.

--- a/dev-requirements.txt
+++ b/dev-requirements.txt
@@ -46,7 +46,8 @@ yamlloader>=0.5.5
 backports.statistics>=0.1.0; python_version == '2.7'
 
 # Virtualenv
-virtualenv>=14.0.0; python_version < '3.8'
+virtualenv>=14.0.0; python_version < '3.5'
+virtualenv>=16.1.0; python_version >= '3.5' and python_version < '3.8'
 virtualenv>=20.0.0; python_version >= '3.8'
 
 # Coverage reporting (no imports, invoked via coveralls script):

--- a/docs/changes.rst
+++ b/docs/changes.rst
@@ -405,6 +405,14 @@ Released: not yet
 * Increased minimum version of 'setuptools' on Python 3.7 from 33.1.1 to 38.4.1
   to fix a bug with new format of .pyc files. (See issue #2167).
 
+* Test: Fixed virtualenv related failures during install test.
+  (See issue #2174)
+
+* Dev: Increased the versions of the base packages 'pip', 'setuptools' and
+  'wheel' to the content of Ubuntu 18.04 as a minimum, and to the lowest
+  versions that support a particular Python versions beyond that.
+  This only affects development of pywbem. (See issue #2174)
+
 **Enhancements:**
 
 * Changed GetCentralInstances methodology in WBEMServer.get_central_instances()

--- a/makefile
+++ b/makefile
@@ -693,7 +693,7 @@ installtest: $(bdist_file) $(sdist_file) tests/installtest/test_install.sh
 ifeq ($(PLATFORM),Windows_native)
 	@echo "makefile: Warning: Skipping install test on native Windows" >&2
 else
-	tests/installtest/test_install.sh $(bdist_file) $(sdist_file)
+	tests/installtest/test_install.sh $(bdist_file) $(sdist_file) $(PYTHON_CMD)
 endif
 	@echo "makefile: Done running install tests"
 

--- a/minimum-constraints.txt
+++ b/minimum-constraints.txt
@@ -41,18 +41,46 @@
 #   pip         19.3.1
 #   setuptools  41.6.0
 #   wheel       0.33.6
+#
+# Info: Python version supported by base packages (added / removed)
+# * Python 2.7 support:
+#   pip         ?
+#   setuptools  ? / 45.0.0
+#   wheel       ?
+# * Python 3.4 support:
+#   pip         ? / 19.2
+#   setuptools  ? / 44.0.0
+#   wheel       ? / 0.34.0
+# * Python 3.5 support:
+#   pip         8.0
+#   setuptools  18.3
+#   wheel       0.30.0
+# * Python 3.6 support:
+#   pip         10.0.0
+#   setuptools  34.0.0
+#   wheel       0.30.0
+# * Python 3.7 support:
+#   pip         18.1
+#   setuptools  40.6.0
+#   wheel       0.32.0
+# * Python 3.8 support:
+#   pip         19.3.1
+#   setuptools  41.5.0
+#   wheel       0.33.5
 
-# For Python up to Python 3.7, we use the versions from Ubuntu 17.04:
-# setuptools 38.4.1 fixes a bug on Python 3.7.
-pip==9.0.1; python_version <= '3.7'
-setuptools==33.1.1; python_version < '3.7'
-setuptools==38.4.1; python_version == '3.7'
-wheel==0.29.0; python_version <= '3.7'
-
-# For Python 3.8 and above, we use the versions from Python 3.8.0:
+# For the base packages, we use the versions from Ubuntu 18.04 as a general
+# minimum, and then increase it to the first version that introduced support
+# for a particular Python version:
+pip==9.0.1; python_version <= '3.5'
+pip==10.0.0; python_version == '3.6'
+pip==18.1; python_version == '3.7'
 pip==19.3.1; python_version >= '3.8'
-setuptools==41.6.0; python_version >= '3.8'
-wheel==0.33.6; python_version >= '3.8'
+setuptools==39.0.1; python_version <= '3.6'
+setuptools==40.6.0; python_version == '3.7'
+setuptools==41.5.0; python_version >= '3.8'
+wheel==0.30.0; python_version <= '3.6'
+wheel==0.32.0; python_version == '3.7'
+wheel==0.33.5; python_version >= '3.8'
 
 
 # Direct dependencies for install (must be consistent with requirements.txt)
@@ -90,7 +118,8 @@ importlib-metadata==0.12; python_version <= '3.7'
 pytz==2016.10
 
 # Virtualenv
-virtualenv==14.0.0; python_version < '3.8'
+virtualenv==14.0.0; python_version < '3.5'
+virtualenv==16.1.0; python_version >= '3.5' and python_version < '3.8'
 virtualenv==20.0.0; python_version >= '3.8'  # requires six<2,>=1.12.0
 
 # Unit test (indirect dependencies):

--- a/tests/installtest/test_install.sh
+++ b/tests/installtest/test_install.sh
@@ -34,10 +34,6 @@ VERBOSE="true"
 MYNAME=$(basename "$0")
 MYDIR=$(dirname "$0")    # Directory of this script, as seen by caller
 
-# Python command for creating virtualenvs.
-# Using just "python" allows using any active virtualenv.
-PYTHON="python"
-
 # Repo root dir, as seen by caller
 # Using MYDIR makes the script run with any caller's CWD.
 ROOT_DIR="$MYDIR/../.."
@@ -147,19 +143,27 @@ function make_virtualenv()
     remove_virtualenv $envname
   fi
 
-  python_cmd=$(which $PYTHON)
+  python_cmd_path=$(which $PYTHON_CMD)
   verbose "Creating virtualenv: $envdir"
-  verbose "..using this Python command: $python_cmd"
-  run "virtualenv -p $python_cmd $envdir"
+  verbose "..using this Python command: $python_cmd_path"
+  verbose "..that has this Python version: $($PYTHON_CMD --version 2>&1)"
+  run "virtualenv -p $python_cmd_path $envdir"
 
-  verbose "Activating virtualenv: $envdir"
-  run "source $envdir/bin/activate"
-  echo "Python command in this virtualenv: $(which python)"
-  echo "Python version in this virtualenv: $(python --version 2>&1)"
+  run "source $envdir/bin/activate" "Activating virtualenv: $envdir"
+  verbose "Python command in this virtualenv: $(which python)"
+  verbose "Python version in this virtualenv: $(python --version 2>&1)"
 
-  verbose "Installing pip, setuptools, wheel with PACKAGE_LEVEL=$PACKAGE_LEVEL"
-  run "pip install pip $PIP_OPTS"
-  run "pip install setuptools wheel $PIP_OPTS"
+  verbose "Packages in this virtualenv before reinstalling base packages:"
+  pip list --format=columns 2>/dev/null || pip list 2>/dev/null
+
+  # If pip is reinstalled (=downgraded) before reinstalling setuptools and wheel,
+  # we get permission errors on Travis when running with minimum package levels.
+  run "pip install setuptools $PIP_OPTS" "Reinstalling setuptools with PACKAGE_LEVEL=$PACKAGE_LEVEL"
+  run "pip install wheel $PIP_OPTS" "Reinstalling wheel with PACKAGE_LEVEL=$PACKAGE_LEVEL"
+  run "pip install pip $PIP_OPTS" "Reinstalling pip with PACKAGE_LEVEL=$PACKAGE_LEVEL"
+
+  verbose "Packages in this virtualenv before actual install test:"
+  pip list --format=columns 2>/dev/null || pip list 2>/dev/null
 }
 
 function remove_virtualenv()
@@ -508,17 +512,22 @@ function test5()
 
 WHL_DISTFILE="$1"  # absolute or relative to caller's cwd
 SRC_DISTFILE="$2"  # absolute or relative to caller's cwd
+PYTHON_CMD="$3"    # Python command to use (outside of the created virtualenvs)
 
-if [[ -z $SRC_DISTFILE ]]; then
-  error "Arguments missing. Usage: $MYNAME WHEEL_DIST_FILE SOURCE_DIST_FILE"
+if [[ -z $PYTHON_CMD ]]; then
+  error "Arguments missing. Usage: $MYNAME WHEEL_DIST_FILE SOURCE_DIST_FILE PYTHON_CMD"
   exit 2
+fi
+if [[ ! -f $WHL_DISTFILE ]]; then
+  error "Wheel distribution archive does not exist: $WHL_DISTFILE"
+  exit 1
 fi
 if [[ ! -f $SRC_DISTFILE ]]; then
   error "Source distribution archive does not exist: $SRC_DISTFILE"
   exit 1
 fi
-if [[ ! -f $WHL_DISTFILE ]]; then
-  error "Wheel distribution archive does not exist: $WHL_DISTFILE"
+if ! which $PYTHON_CMD >/dev/null 2>&1; then
+  error "Cannot find Python command: $PYTHON_CMD"
   exit 1
 fi
 

